### DISCUSSION TITLE
3D Panel can't load meshes from URDF file

### DISCUSSION THREAD

**Discussion by RobinHeitz** [foxglove/community#627] [Q&A]
Hi there, I'm trying for hours to getting this to work but can't figure it out.

I've a ros2 package with a robot description and a launch file, that launches 

1. foxglove_bridge
2. robot_state_publisher
3. joint_state_publisher

In Foxglove Studio (Local App), I can see the links coordinate systems, but no meshes. There is an error message within the panel settings: /robot_description ->

`Error loading mesh from "package://rebel_test/meshes/Robolink_RebeL_001.dae": Failed to fetch: package://rebel_test/meshes/Robolink_RebeL_001.dae
`

**URDF File (short for testing purpose):**
```
<?xml version="1.0"?>
<robot name="igus_rebel" >

  <link name="world" />

  <link name="base_link">
    <visual>
      <geometry>
        <mesh filename="package://rebel_test/meshes/Robolink_RebeL_000.dae"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://rebel_test/meshes/Robolink_RebeL_000.dae"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="1.0" />
      <inertia ixx="1.0" iyy="1.0" izz="1.0" ixy="0.0" ixz="0.0" iyz="0.0" />
    </inertial>
  </link>

  <joint name="world_joint" type="fixed">
    <parent link="world" />
    <child link = "base_link" />
    <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0" />
  </joint>

</robot>
```

**My display.launch.py file:**
```
import launch
from launch.substitutions import Command, LaunchConfiguration
import launch_ros
import os

def generate_launch_description():
    pkg_share = launch_ros.substitutions.FindPackageShare(package='rebel_test').find('rebel_test')
    default_model_path = os.path.join(pkg_share, 'urdf/rebel.urdf')

    robot_state_publisher_node = launch_ros.actions.Node(
        package='robot_state_publisher',
        executable='robot_state_publisher',
        parameters=[{'robot_description': Command(['xacro ', LaunchConfiguration('model')])}]
    )
    
    joint_state_publisher_node = launch_ros.actions.Node(
        package='joint_state_publisher',
        executable='joint_state_publisher',
        name='joint_state_publisher',
    )

    foxglove_node = launch_ros.actions.Node(
        package='foxglove_bridge',
        executable='foxglove_bridge',
        name='foxglove_bridge',
    )

    return launch.LaunchDescription([
        launch.actions.DeclareLaunchArgument(name='model', default_value=default_model_path, description='Absolute path to robot urdf file'),
        joint_state_publisher_node,
        robot_state_publisher_node,
        foxglove_node,
    ])
```

I dont't know what to try anymore :/ I read about foxglove doesn't support relative paths within urdf files, but thats not the case here.
Thanks in advance!

**Comment by RobinHeitz** [comment]
Meanwhile, I tried both, the Foxglove App (with foxglove_node) and the chrome app (using Websockets and rosbridge).
If i set the geometry to "objects" like cylinder etc. they get visible within the 3D-panel.

I tried all combinations of relative paths within URDF Files, and of course the recommended way with "package://path" and this does not work. If i go for relative paths, the error message changes to 404.

I checked, that the desired files are within the `pkg-name/share/package-name/meshes/my_mesh.dae `directory.

Dunno what else to try :(

**Reply by defunctzombie** [reply]
From your initial post it sounds to me like the Foxglove Studio app and your ROS workspace are on different hosts or in different environments? The URDF `package://` path lookup only works if Foxglove Studio is running on the same computer (no containers, separate hosts, etc) as your ROS workspace. This is because the app actually has to go and find the files on your local disk relative to the ament prefix path.

Can you confirm where your ROS workspace is relative to where you are running Foxglove Studio please.

**Reply by RobinHeitz** [reply]
Well you're right, I'm having my WS in the devcontainer...
I read about foxglove from somebody else on answers.ros, in the context with docker, I assumed he was running foxglove + docker containers...

So there is probably no way to combine both things, right?
Thank you a lot for the answer! Now I need to check how I can get humble running native on my mac.

**Reply by defunctzombie** [reply]
Generally there is not a way to have Foxglove Studio read the files in your container. I say generally because there's all sorts of volume mounting things folks sometimes do to have the container files also on the local file system. What you can try is to put these files `rebel_test/meshes/Robolink_RebeL_000.dae` on your local machine and then configure the `ROS_PACKAGE_PATH` in Studio settings (the settings cog in the lower right corner) to point at the directory containing the `rebel_test` folder. That _might_ work and you don't need a full ROS workspace on your host, only something that looks like one that Studio can traverse to find the mesh files that you want to find with the `package://` urls.

**Reply by RobinHeitz** [reply]
I set the ROS_PACKAGE_PATH like you said and its working, but only in the App (not the chrome web app) - but thats more than enough for me 👍  

I'm actually using a devcontainer setup, so I don't actually need to copy files or mount any directory, because the files are already there and accessible locally.

Thank you !
